File: tools/changelog-generator/changelog.py
import requests
import os
from datetime import datetime, timedelta


base_url = 'https://api.github.com'
repo = 'Kong/docs.konghq.com'
current_directory = os.getcwd()
file_path = os.path.join(current_directory, 'docs_changelog.md')


today = datetime.now()
last_week = today - timedelta(days=7)

since = last_week.date().isoformat()
until = today.date().isoformat()

#api_key = 'insert API key'
api_key = os.getenv('API_KEY')

pr_url = f'{base_url}/repos/{repo}/pulls?state=closed&base=main&sort=updated&direction=desc&since={since}&until={until}'


headers = {
    'Authorization': f'Token {api_key}'
}


pr_response = requests.get(pr_url, headers=headers)


if pr_response.status_code == 200:
    prs = pr_response.json()

    
    markdown_content = ""

   
    current_week_title = f"## {last_week.strftime('%Y-%m-%d')} - {today.strftime('%Y-%m-%d')}"
    markdown_content += current_week_title + "\n\n"

    
    for pr in prs:
        pr_number = pr['number']
        pr_title = pr['title']
        pr_user = pr['user']['login']
        pr_url = pr['html_url']
        commit_sha = pr['merge_commit_sha']

        
        commit_url = f'{base_url}/repos/{repo}/commits/{commit_sha}'

        
        commit_response = requests.get(commit_url, headers=headers)

        
        if commit_response.status_code == 200:
            commit = commit_response.json()
            #commit_name = pr_title

            
            files = commit['files']
            new_files = [file['filename'] for file in files if file['status'] == 'added' and file['filename'].startswith('app/_src/')]
            updated_files = [file['filename'] for file in files if file['status'] == 'modified' and file['filename'].startswith('app/_src/')]

            
            markdown_content += f"### PR #{pr_number}: [{pr_title}]({pr_url})\n"
            #markdown_content += f"- Commit: {commit_name}\n"

            
            if new_files:
                markdown_content += "\n**New**\n\n"
                for file in new_files:
                    parts = file.split('/')
                    product_name = parts[2] 
                    url = f"https://docs.konghq.com/{product_name}/latest/{'/'.join(parts[3:]).replace('.md', '')}"
                    markdown_content += f"- [{file}]({url})\n"

            
            if updated_files:
                markdown_content += "\n**Updates**\n\n"
                for file in updated_files:
                    parts = file.split('/')
                    product_name = parts[2] 
                    url = f"https://docs.konghq.com/{product_name}/latest/{'/'.join(parts[3:]).replace('.md', '')}"
                    markdown_content += f"- [{file}]({url})\n"

            markdown_content += "\n---\n\n"
        else:
            print(f'Error: Failed to get commit details. Status code: {commit_response.status_code}')
else:
    print(f'Error: Failed to get pull requests. Status code: {pr_response.status_code}')


with open('docs_changelog.md', 'r') as file:
    existing_content = file.read()


new_content = markdown_content + existing_content


with open(file_path, 'w') as file:
    file.write(new_content)

File: tools/changelog-generator/changelog.js
const fs = require("fs").promises;
const { subDays, startOfDay, format } = require("date-fns");
const groupBy = require("lodash.groupby");

const convertFilePathsToUrls = require("../_utilities/path-to-url");
const now = startOfDay(new Date());
const earliestDate = subDays(now, 7);

(async function () {
  // Create an octokit instance
  const { Octokit } = require("@octokit/rest");
  const octokit = new Octokit({
    auth: process.env.GITHUB_TOKEN,
  });

  const owner = "Kong";
  const repo = "docs.konghq.com";

  let pulls = await octokit.paginate(
    octokit.pulls.list,
    {
      owner,
      repo,
      state: "closed",
      per_page: 100,
    },
    (response, done) => {
      const validPulls = [];

      for (let pr of response.data) {
        //console.log(JSON.stringify(pr));
        //process.exit(0);
        prDate = new Date(pr.merged_at);
        if (prDate > earliestDate && pr.merged_at && pr.base.ref == "main") {
          validPulls.push(pr);
        }
      }

      // If the last prDate we read is < earliestDate then stop fetching
      if (prDate < earliestDate) {
        done();
      }
      return validPulls;
    },
  );

  //pulls = pulls.slice(0, 1);

  let prsWithFiles = (
    await Promise.all(
      pulls.map(async (pr) => {
        const created_at = new Date(pr.merged_at);



        // Load the files for this PR using the API
        const pr_files = await octokit.paginate(
          octokit.rest.pulls.listFiles,
          {
            owner,
            repo,
            pull_number: pr.number,
          },
          (response) => response.data,
        );

        // Convert app, _src, /hub/ etc paths to URLs
        affected_files = (
          await convertFilePathsToUrls(pr_files, { skip_nav_files: true })
        ).map((v) => {
          return {
            url: `https://docs.konghq.com${v.url}`,
            status: v.status,
          };
        });

        return {
          number: pr.number,
          title: pr.title,
          description: extractDescription(pr.body),
          user: pr.user.login,
          url: pr.html_url,
          labels: pr.labels,
          created_at,
          week: format(created_at, "I"),
          affected_files,
        };
      }),
    )
  )
    
    .filter((pr) => pr.affected_files.length > 0)
    .filter((pr) => !pr.labels.map(l => l.name).includes("skip-changelog"));
  prsWithFiles = groupBy(prsWithFiles, "week");

  // Write the changelog
  let changelogContent = `# Changelog\n\n<!--vale off-->\n\n`;

  for (let week of Object.keys(prsWithFiles).reverse()) {
    changelogContent += `## Week ${week}\n\n`;

    for (const pr of prsWithFiles[week]) {
      changelogContent += `### [${pr.title}](${pr.url}) (${format(
        pr.created_at,
        "yyyy-MM-dd",
      )})\n\n`;

      changelogContent += `${pr.description}\n\n`;

      const addedFiles = pr.affected_files.filter((u) => u.status == "added");
      const changedFiles = pr.affected_files.filter(
        (u) => u.status == "modified",
      );

      if (addedFiles.length) {
        changelogContent += `#### Added\n\n`;
        for (const file of addedFiles) {
          changelogContent += `- ${file.url}\n`;
        }
      }

      if (addedFiles.length && changedFiles.length) {
        changelogContent += `\n`;
      }

      if (changedFiles.length) {
        changelogContent += `#### Modified\n\n`;
        for (const file of changedFiles) {
          changelogContent += `- ${file.url}\n`;
        }
      }

      changelogContent += `\n\n`;
    }
  }

  changelogContent = changelogContent.trim();

  // Prepend changelogContent to changelog.md
  const changelogFile = __dirname + "/../../changelog.md";
  let existingChangelog = "";

  try {
    existingChangelog = await fs.readFile(changelogFile, "utf8");
  } catch (e) {
    console.log(e); // for debugging why the file is being overwritten.
  }

  changelogContent = `${changelogContent}\n\n${existingChangelog.replace(
    "# Changelog\n\n<!--vale off-->\n\n",
    "",
  )}`;
  await fs.writeFile(changelogFile, changelogContent);
})();

function extractDescription(body) {
  body = body || "";
  return body
    .split("### Testing instructions")[0]
    .replace("### Description", "")
    .trim();
}
